### PR TITLE
check view consistency in fetch/exercise by interface

### PR DESCRIPTION
**Pull request by paulbrauner-da**
Addresses the runtime part of UP-1 in [this doc](https://docs.google.com/spreadsheets/d/1_0ER--x-YJzoYSoy7VFs6emzn4ZYjBc0kte4MLosuxs/edit?gid=0#gid=0).

Make sure to "hide whitespace" if reviewing on github: quite a lot of code is untouched but indented at a different level.

### CODE REVIEW

File: sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -12,6 +13,7 @@ import com.daml.lf.data._
 import com.daml.lf.data.Numeric.Scale
 import com.daml.lf.interpretation.{Error => IE}
 import com.daml.lf.language.Ast
+import com.daml.lf.language.Ast.TTyCon
```
Comment (remyhaemmerle-da) [inline sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:16]:
```suggestion
```

File: sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 2]
```diff
@@ -1139,41 +1141,198 @@ private[lf] object SBuiltin {
...
+      machine: UpdateMachine,
+      coid: V.ContractId,
+      ifaceId: TypeConName,
+  )(k: SAny => Control[Question.Update]): Control[Question.Update] = {
+    fetchAny(machine, None, coid, SValue.SValue.None) { (_, srcContract) =>
+      val (tplId, arg) = getSAnyContract(ArrayList.single(srcContract), 0)
+      castAnyInterface(machine, ifaceId, coid, tplId, arg) { _ =>
+        viewInterface(machine, ifaceId, tplId, arg) { srcView =>
+          executeExpression(machine, SEPreventCatch(srcView)) { _ =>
+            k(SAny(TTyCon(tplId), arg))
```
Comment (remyhaemmerle-da) [inline sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1203]:
```suggestion
            k(SAny(Ast.TTyCon(tplId), arg))
```

File: sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 3]
```diff
@@ -1139,41 +1141,198 @@ private[lf] object SBuiltin {
...
+      machine: Machine[_],
+      tplId: TypeConName,
+      ifaceId: TypeConName,
+  )(k: () => Control[Q]): Control[Q] = {
+    val tplImplementsInterface =
+      machine.compiledPackages.pkgInterface
+        .lookupTemplate(tplId)
+        .exists(_.implements.contains(ifaceId))
+    if (tplImplementsInterface) k()
+    else crash(s"template $tplId does not implement interface $ifaceId")
```
Comment (remyhaemmerle-da) [inline sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1318]: Should this be a crash ?

Comment (remyhaemmerle-da) [reply sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1318]: if yes, please add a comment.

File: sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 4]
```diff
@@ -1428,19 +1597,28 @@ private[lf] object SBuiltin {
...
+      record: SValue,
+  )(k: SExpr => Control[Q]): Control[Q] = {
+    val ref = getInterfaceInstance(machine, ifaceId, templateId).fold(
+      crash(
+        s"Attempted to call view for interface ${ifaceId} on a wrapped " +
+          s"template of type ${ifaceId}, but there's no matching interface instance."
+      )
+    )(iiRef => InterfaceInstanceViewDefRef(iiRef))
+    val e = SEApp(SEVal(ref), Array(record))
+    k(e)
```
Comment (remyhaemmerle-da) [inline sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1619]:
```suggestion
    k(SEApp(SEVal(ref), Array(record)))
```

**Review by remyhaemmerle-da — APPROVED** [review]
Nice.

File: sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 3 above)
Comment (paulbrauner-da) [reply sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala]: Turns out `castAnyInterface` already does that check and its result is never used. So I:
- deleted this flawed version of `ensureTemplateImplementsInterface`
- slightly modified `castAnyInterface` to return no result 
- renamed  `castAnyInterface` to `ensureTemplateImplementsInterface`
- deleted the `SBCastAnyInterface` builtin that was no longer used

File: sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 4 above)
Comment (paulbrauner-da) [reply sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala]: Done

**Pull request merged**